<div style="display: flex; align-items: center; gap: 20px; font-family: Arial, sans-serif;">
    <div style="margin-right: 30px;">
        <img src="../Sources/Logo.png" width="300" style="border-radius: 10px;"/>
    </div>
    <div style="line-height: 1.5;">
        <strong style="font-size: 18px; display: block; margin-bottom: 5px;">
            <a href='https://t.me/M4m4ev_Alex' style="text-decoration: none; color: #1a73e8;">
                Репетитор по информатике | Мамаев Алексей
            </a>
        </strong>
        <span style="font-size: 16px; color: #333;">
            Разбор <a href='https://euler.jakumo.org/problem/7-10001-e-prostoe-cislo' style="text-decoration: none; color: #1a73e8;">
            7-й задачи</a> проекта Эйлера
        </span>
        <span style="font-size: 16px; color: #333;">
            <br><br><br><br><br><br>ТГК: <a href='https://t.me/M4m4ev_Alex' style="text-decoration: none; color: #1a73e8;">@M4m4ev_Alex</a><br>
        </span>
    </div>
</div>

### Задача 7

Первые шесть простых чисел:

```text
2, 3, 5, 7, 11, 13
```

Шестым простым числом является `13`.

Необходимо определить, какое число является **10001-м простым числом**.

Ответ: 104743

## 1) Последовательный поиск простых чисел

В данном решении простые числа находятся последовательно, начиная с числа `2`.

Для проверки каждого кандидата используется собственная функция `is_prime()`, которая определяет, является ли число простым.

Поиск продолжается до тех пор, пока не будет найдено **10001-е простое число**.

### Алгоритм

1. Создаётся функция `is_prime()`, проверяющая число на простоту.
2. Счётчик найденных простых чисел инициализируется значением `2`, так как числа `2` и `3` уже являются простыми.
3. Далее последовательно перебираются только нечётные числа.
4. Для каждого числа выполняется проверка на простоту.
5. Если число является простым, счётчик увеличивается.
6. После нахождения 10001-го простого числа оно возвращается в качестве ответа.

### Особенности

- Используется собственная реализация проверки числа на простоту.
- Проверяются только нечётные числа, что уменьшает количество вычислений почти в два раза.
- Делители проверяются только до `√n`, поскольку большего делителя без соответствующей пары существовать не может.
- Временная сложность проверки одного числа составляет `O(√n)`.
- Решение не требует дополнительной памяти.

### Вывод

Последовательный перебор простых чисел является достаточно простым и наглядным способом решения задачи. Несмотря на отсутствие сложных алгоритмов, использование проверки делителей только до квадратного корня и пропуск чётных чисел значительно ускоряют поиск.

In [8]:
from time import perf_counter # измерение времени выполнения алгоритмов

In [9]:
def is_prime(num: int) -> bool:
    if num <= 1:
        return False
    if num == 2:
        return True
    if num % 2 == 0:
        return False

    for i in range(3, int(num ** 0.5) + 1, 2):
        if num % i == 0:
            return False
    return True


def get_result_1(num: int) -> int:
    c = 2
    n = 3
    while c != num:
        n += 2
        if is_prime(n):
            c += 1
    return n


start = perf_counter()
print(get_result_1(10_001))
end = perf_counter()

print(f'{end - start:.5f}')

104743
0.09683


## 2) Использование встроенной функции `isprime()`

В данном решении используется тот же алгоритм последовательного поиска простых чисел, однако проверка числа на простоту выполняется с помощью встроенной функции `isprime()` из библиотеки `sympy`.

### Особенности

- Используется готовая функция `isprime()` из библиотеки `sympy`.
- Код становится значительно компактнее.
- Встроенная реализация оптимизирована и обычно работает быстрее собственной.

### Вывод

Использование `isprime()` позволяет существенно упростить код, сохранив прежнюю идею решения задачи.

In [10]:
from sympy import isprime # функция для проверки числа на простоту

def get_result_2(num: int) -> int:
    c = 2
    n = 3
    while c != num:
        n += 2
        if isprime(n):
            c += 1
    return n

start = perf_counter()
print(get_result_2(10_001))
end = perf_counter()

print(f'{end - start:.5f}')

104743
0.08873


## 3) Решето Эратосфена

В предыдущих решениях каждое число последовательно проверялось на простоту.

В данном подходе используется **решето Эратосфена** — один из наиболее эффективных алгоритмов поиска простых чисел в заданном диапазоне.

Основная идея заключается в том, что вместо проверки каждого числа на простоту по отдельности алгоритм последовательно исключает все составные числа, оставляя только простые.

### Верхняя граница поиска

Для построения решета необходимо заранее определить диапазон, в котором гарантированно находится искомое простое число.

Для этого используется оценка

$$
p_n < n\left(\ln n + \ln \ln n\right),
\qquad n \ge 6
$$

где:

- $p_n$ — `n`-е простое число;
- $\ln$ — натуральный логарифм.

В программе верхняя граница вычисляется следующим образом:

```python
limit = int(n * (log(n) + log(log(n)))) + 1
```

Полученное значение является верхней оценкой, поэтому решето гарантированно будет содержать искомое простое число.

---

### Алгоритм

1. Вычисляется верхняя граница поиска `limit`.
2. Создаётся массив, в котором первоначально все числа считаются простыми.
3. Последовательно перебираются числа, начиная с `2`.
4. Для каждого найденного простого числа все его кратные помечаются как составные.
5. После завершения работы алгоритма остаются только простые числа.
6. Возвращается `n`-е простое число.

### Особенности

- Каждое составное число исключается только один раз.
- Не требуется отдельно проверять каждое число на простоту.
- Временная сложность составляет `O(n log log n)`.
- Пространственная сложность составляет `O(n)`.
- Является одним из наиболее эффективных алгоритмов поиска большого количества простых чисел.

### Вывод

Решето Эратосфена значительно превосходит последовательную проверку чисел на простоту при поиске большого количества простых чисел. Благодаря предварительному исключению составных чисел алгоритм позволяет получить результат существенно быстрее, несмотря на использование дополнительной памяти.

In [11]:
from math import log

def get_result_3(n: int) -> int:
    if n == 1:
        return 2

    # Верхняя оценка для n-го простого числа
    limit = int(n * (log(n) + log(log(n)))) + 1

    is_prime = [True] * (limit + 1)
    is_prime[0] = is_prime[1] = False

    for num in range(2, int(limit ** 0.5) + 1):
        if is_prime[num]:
            for multiple in range(num * num, limit + 1, num):
                is_prime[multiple] = False

    primes = [num for num, prime in enumerate(is_prime) if prime]

    return primes[n - 1]


start = perf_counter()
print(get_result_3(10_001))
end = perf_counter()

print(f'{end - start:.5f}')

104743
0.01241


## Сравнение времени выполнения

Для сравнения производительности используется функция `timeit()`, предназначенная специально для измерения времени выполнения небольших участков кода.

Каждый алгоритм запускается **100 раз**, после чего общее время делится на количество запусков. Таким образом получается **среднее время выполнения одного запуска**.

In [14]:
from timeit import timeit

funcs = (
    get_result_1,
    get_result_2,
    get_result_3,
)

for i, func in enumerate(funcs, start=1):
    tm = timeit(lambda: func(10_001), number=100) / 100
    print(f'Алгоритм {i}: {tm:.5f}')

Алгоритм 1: 0.07290
Алгоритм 2: 0.05623
Алгоритм 3: 0.01157


## Заключение

В ходе решения задачи были рассмотрены три различных подхода к поиску **10001-го простого числа**.

Первый алгоритм основан на последовательном переборе нечётных чисел с использованием собственной функции проверки на простоту. Благодаря проверке делителей только до квадратного корня и пропуску чётных чисел удалось значительно сократить количество вычислений, однако каждое число по-прежнему приходится проверять отдельно.

Во втором решении используется встроенная функция `isprime()` из библиотеки `sympy`. Основная идея алгоритма остаётся прежней, однако применение оптимизированной реализации проверки на простоту позволило уменьшить время выполнения и сделать код значительно компактнее.

Наиболее эффективным оказалось третье решение, использующее **решето Эратосфена**. Вместо проверки каждого числа на простоту алгоритм заранее исключает все составные числа, благодаря чему поиск простых чисел выполняется значительно быстрее.

Среднее время выполнения алгоритмов:

- **Алгоритм 1:** `0.07290` с
- **Алгоритм 2:** `0.05623` с
- **Алгоритм 3:** `0.01157` с

Полученные результаты показывают, что при поиске большого количества простых чисел использование решета Эратосфена значительно эффективнее последовательной проверки каждого кандидата. В данном случае третье решение оказалось более чем в **6 раз быстрее** первой реализации.